Add JIT support for IN timestamp expressions

zebs97 opened the pull request.

zebs97 commented on core/src/test/java/io/questdb/test/jit/CompiledFilterIRSerializerTest.java:
```diff
@@ -847,9 +847,10 @@ public void testInSubSelect() throws Exception {
+    @Test
     public void testTimestampInLiteral() throws Exception {
         serialize("atimestamp in '2020-01-01'");
+        assertIR("(i64 1577836800000000L)(i64 atimestamp)(>=)(i64 1577923199999999L)(i64 atimestamp)(<=)(&&)(ret)");
```
I'll add more tests using more complex range expressions

puzpuzpuz replied:
Yes, more tests are certainly required, including tests that verify that we get the same query results with and without enabled JIT.

puzpuzpuz replied:
Also, advanced intervals should be tested, e.g. `ts in '2023-12-20;15s'` or `ts in '2023-09-20T23:59:58;4s;-1d;7'`

zebs97 replied:
Done. Added IR tests for range expressions as well as a results test with and without JIT in `ParallelFilterTest`

puzpuzpuz commented on core/src/test/java/io/questdb/test/jit/CompiledFilterIRSerializerTest.java:
```diff
@@ -847,9 +847,16 @@ public void testInSubSelect() throws Exception {
+        assertIR("(i64 1577836800000000L)(i64 atimestamp)(>=)(i64 1577923199999999L)(i64 atimestamp)(<=)(&&)(ret)");
+        serialize("atimestamp in '2020-01-01;15s'");
+        assertIR("(i64 1577836800000000L)(i64 atimestamp)(>=)(i64 1577923214999999L)(i64 atimestamp)(<=)(&&)(ret)");
+        serialize("atimestamp in '2020-01-01T23:59:58;4s;-1d;3'");
```
Would be great to add a test with a more complex filter involving other predicates, e.g. `along = 42 or atimestamp in '2020-01-01T23:59:58;4s;-1d;3'`

zebs97 replied:
Added in last commit

puzpuzpuz commented on core/src/test/java/io/questdb/test/griffin/ParallelFilterTest.java:
```diff
@@ -797,4 +807,47 @@ private void testIn(int jitMode) throws Exception {
+                    TestUtils.assertSql(
+                            engine,
+                            sqlExecutionContext,
+                            "select * from tab where ts in '1970-01-01T00:00:00;3m;1d;5'",
```
Let's use a more complex filter involving multiple predicates:
```suggestion
                            "select * from tab where ts in '1970-01-01T00:00:00;3m;1d;5' and value = 't3'",
```

ideoma replied:
The designated timestamp filter of this simple form will be processed separately from WHERE clause in `WhereClauseParser`. To test the code you added the column in the filter should be non-designated timestamp

zebs97 replied:
Added additional filter and changed the timestamp filter variable to a new one `preciseTs` which is not the designated timestamp

puzpuzpuz commented on core/src/main/java/io/questdb/jit/CompiledFilterIRSerializer.java:
```diff
@@ -1029,6 +1034,39 @@ private void serializeIn() throws SqlException {
+
+        if (intervalEx == null) return;
+
+        final LongList intervals = new LongList();
```
Please introduce a new field for the interval list and reuse it between JIT compilations. There is `LongList#clear()` which resets the list.

zebs97 replied:
Done in last commit

puzpuzpuz commented on core/src/test/java/io/questdb/test/jit/CompiledFilterIRSerializerTest.java:
```diff
@@ -847,9 +847,16 @@ public void testInSubSelect() throws Exception {
 
-    @Test(expected = SqlException.class)
+    @Test
     public void testTimestampInLiteral() throws Exception {
```
Please add a test for `atimestamp in null`

zebs97 replied:
What's the expected output of such case? Evaluate to `false` or throw a `SqlException`?

puzpuzpuz replied:
It should be same as in `IntervalUtils#parseAndApplyIntervalEx()` with null `seq` input. In fact, you could use that method to make sure that the behavior is the same as in `InTimestampStrFunctionFactory`. To do that, you'd need to parse null literal (`tok == null || SqlKeywords.isNullKeyword(tok)`).

zebs97 replied:
Added null handling leveraging `IntervalUtils.parseAndApplyIntervalEx` and unit test for it

puzpuzpuz commented on core/src/test/java/io/questdb/test/jit/CompiledFilterIRSerializerTest.java:
```diff
@@ -847,9 +847,16 @@ public void testInSubSelect() throws Exception {
 
-    @Test(expected = SqlException.class)
+    @Test
     public void testTimestampInLiteral() throws Exception {
```
Please add a test with bind variables, e.g.
```java
        bindVariableService.clear();
        bindVariableService.setStr("str", "2020");
        serialize("atimestamp in :str");
```

Such filters shouldn't be JIT compiled since interval parsing has to be done in runtime.

zebs97 replied:
Done in last commit

puzpuzpuz commented:
Hi @zebs97 

Do you think you can find some time to update the PR and address the review comments anytime soon? If not, no problem at all, we can pick it up and finalize it ourselves.

zebs97 commented:
Hi @puzpuzpuz 

Sorry, I haven’t had much time since I last worked in this. I’ll save some time this week to address the comments, I’d like to complete the work.

Thanks.

puzpuzpuz commented:
Thanks!

zebs97 commented:
Hi @puzpuzpuz I addressed all comments and fixed the merge conflicts, the PR should be ready for review. Thanks for the patience and sorry again for not finishing this work earlier.

puzpuzpuz commented on core/src/main/java/io/questdb/jit/CompiledFilterIRSerializer.java:
```diff
@@ -1029,6 +1034,37 @@ private void serializeIn() throws SqlException {
+        final LongList intervals = predicateContext.inIntervals;
+        IntervalUtils.parseAndApplyIntervalEx(intervalEx, intervals, position);
+
+        final PostOrderTreeTraversalAlgo traverseAlgo = new PostOrderTreeTraversalAlgo();
```
nit: would be great to reuse this object across JIT compilations.

zebs97 replied:
Done, also added a new test `testInAndInTimestamp` to ensure the single `PostOrderTreeTraversalAlgo` handles multiple `in` expressions correctly without one stepping on the other.

puzpuzpuz commented:
@zebs97 there are failing tests in `SqlCompilerImplTest`, `InStrTest`, and `InVarcharTest`. Could you check?

zebs97 commented:
@puzpuzpuz fixed the failing tests. Failures were due to expressions such as `var in null` where var was not a timestamp, which went through the timestamp serialization branch, producing unexpected results.

I added an additional check for the predicate expression type before going through with the serialization.

puzpuzpuz commented on core/src/main/java/io/questdb/jit/CompiledFilterIRSerializer.java:
```diff
@@ -871,8 +872,12 @@ private void serializeNumber(
+            } else if (type == ExpressionNode.SET_OPERATION && isInTimestampPredicate()) {
+                serializeInTimestampRange(position);
+            }
             return;
```
`return;` should be moved into above if branches. That's because we want to fall to the subsequent `throw` if there was no above match.

zebs97 replied:
Ah, true, thanks for the catch.

puzpuzpuz commented:
@zebs97 I left one final minor comment. Could you address it, so that the PR is good to go?

zebs97 commented:
Done @puzpuzpuz

puzpuzpuz reviewed (approved):
Many thanks for one more solid contribution!
